pkg/beacon/relay/entry/entry.go

```diff
@@ -33,26 +39,195 @@ func SignAndSubmit(
+		receiveChannel <- netMessage
+	})
+
+	receivedShares := map[group.MemberIndex]*bn256.G1{
+		signer.MemberID(): selfShare,
+	}
+
+	// Run the message loop until the number of received and valid signature
+	// shares is equal to the honest threshold. Message loop will be also
+	// terminated if an other member submits the result or the relay entry
+	// timeout block is reached.
+	for len(receivedShares) < honestThreshold {
+		select {
+		case netMessage := <-receiveChannel:
+			message, ok := netMessage.Payload().(*SignatureShareMessage)
+			if !ok || group.IsMessageFromSelf(signer.MemberID(), message) {
+				continue
+			}
+
+			share, err := extractShare(
+				message,
+				signer.GroupPublicKeyShares(),
+				previousEntry,
+			)
+			if err != nil {
+				logger.Warningf(
+					"[member:%v] rejecting signature share from "+
+						"member [%v]: [%v]",
+					signer.MemberID(),
+					message.senderID,
+					err,
+				)
+				continue
+			}
+
+			logger.Debugf(
+				"[member:%v] accepting signature share from member [%v]",
+				signer.MemberID(),
+				message.senderID,
+			)
+
+			receivedShares[message.senderID] = share
+		case blockNumber := <-relayEntrySubmittedChannel:
+			logger.Infof(
+				"[member:%v] leaving message loop; "+
+					"relay entry submitted by other member at block [%v]",
+				signer.MemberID(),
+				blockNumber,
+			)
+			return nil
+		case blockNumber := <-relayEntryTimeoutChannel:
+			return fmt.Errorf(
+				"relay entry timed out at block [%v]",
+				blockNumber,
+			)
+		}
 	}
 
-	stateMachine := state.NewMachine(channel, blockCounter, initialState)
-	_, _, err = stateMachine.Execute(startBlockHeight)
+	signature, err := completeSignature(signer, receivedShares, honestThreshold)
+	if err != nil {
+		return err
+	}
+
+	submitter := &relayEntrySubmitter{
+		chain:        relayChain,
+		blockCounter: blockCounter,
+		index:        signer.MemberID(),
+	}
+
+	return submitter.submitRelayEntry(
+		signature.Marshal(),
+		signer.GroupPublicKeyBytes(),
+		startBlockHeight,
+		relayEntrySubmittedChannel,
+		relayEntryTimeoutChannel,
+	)
+}
+
+func broadcastShare(
+	ctx context.Context,
+	memberID group.MemberIndex,
+	share *bn256.G1,
+	channel net.BroadcastChannel,
+) {
+	message := &SignatureShareMessage{
+		memberID,
+		share.Marshal(),
+	}
+
+	if err := channel.Send(ctx, message); err != nil {
+		logger.Errorf(
+			"[member:%v] could not send signature share: [%v]",
+			memberID,
+			err,
+		)
+	}
+}
+
+func extractShare(
+	message *SignatureShareMessage,
+	groupPublicKeyShares map[group.MemberIndex]*bn256.G2,
+	previousEntry *bn256.G1,
+) (*bn256.G1, error) {
+	share := new(bn256.G1)
+	_, err := share.Unmarshal(message.shareBytes)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"could not unmarshal signature share: [%v]",
+			err,
+		)
+	}
+
+	publicKeyShare, ok := groupPublicKeyShares[message.senderID]
+	if !ok {
+		return nil, fmt.Errorf(
+			"could not validate signature share; " +
+				"group public key share not found",
```

Comment:
`group public key share for sender not found`? This can help avoid confusion about whose share it is.